File: ios/WalletCore/Coin/Cardano.swift
//
//  Cardano.swift
//  trust-wallet-core-react-native
//
//  Created by Nguyen Van Toan on 21/5/24.
//

import Foundation
import WalletCore

class Cardano : Coin{
    init(){
        super.init(coinType: CoinType.cardano, derivationPath: "m/1852'/1815'/0'/0/0")
    }
    
    override func signTransaction(privateKey : String, tx: Dictionary<String, Any>) throws -> String {
        let utxos: [[String: Any]]? = tx["utxos"] as? [[String: Any]]
        
        guard let utxos = utxos else {
            throw CError.MissingArgumentError
        }
        
        guard let rawPrivateKey = Data(hexString: privateKey) else {
            throw CError.InvalidHexArgumentError
        }
        
        var txs : [CardanoTxInput] = []
        
        for utx in utxos {
            guard let txid = Data(hexString: utx["txid"] as! String) else {
                throw CError.InvalidHexArgumentError
            }
            
            let txInput = CardanoTxInput.with {
                $0.amount = utx["amout"] as! UInt64
                $0.outPoint.txHash = txid
                $0.outPoint.outputIndex = utx["index"] as! UInt64
                $0.address = utx["sender"] as! String
            }
            
            txs.append(txInput)
        }
        
        let input = CardanoSigningInput.with {
            $0.privateKey = [rawPrivateKey]
            $0.ttl = tx["ttl"] as! UInt64
            $0.utxos = txs
            $0.transferMessage.toAddress = tx["to"] as! String
            $0.transferMessage.changeAddress = tx["changeAddress"] as! String
            $0.transferMessage.amount = tx["amount"] as! UInt64
        }
        
        let signed : CardanoSigningOutput = AnySigner.sign(input: input, coin: coinType)
        
        return signed.encoded.hexString
    }
}

File: src/parameter/cardano.ts
import { Parameter } from './parameter';

class CardanoTransaction {
  private amount: Number;
  private txHash: String;
  private index: Number;
  private sender: String;

  constructor(amount: Number, txHash: String, index: Number, sender: String) {
    this.amount = amount;
    this.txHash = txHash;
    this.index = index;
    this.sender = sender;
  }

  toNativeMap(): Object {
    let map = {
      txid: this.txHash,
      amount: this.amount,
      index: this.index,
      sender: this.sender,
    };

    return map;
  }
}

export class CardanoTransactionParameter extends Parameter<
  [CardanoTransaction]
> {
  private toAddress: String;
  private changeAddress: String;
  private amount: Number;
  private ttl: Number;

  constructor(
    toAddress: String,
    changeAddress: String,
    amount: Number,
    ttl: Number,
    transaction: undefined
  ) {
    super(transaction);
    this.toAddress = toAddress;
    this.amount = amount;
    this.changeAddress = changeAddress;
    this.ttl = ttl;
  }

  toNativeMap(): Object {
    let map = {
      utxos: this.getTransaction()?.map((e) => e.toNativeMap()),
      ttl: this.ttl,
      to: this.toAddress,
      changeAddress: this.changeAddress,
      amount: this.amount,
    };

    return map;
  }
}
